donut/modules/voting/routes.py

```diff
@@ -0,0 +1,264 @@
+        flask.flash(restrict_message)
+        return list_surveys()
+
+    survey_id = survey['survey_id']
+    return helpers.process_params_request(True, survey_id, access_key)
+
+
+@blueprint.route('/1/surveys/<access_key>/questions', methods=['POST'])
+def save_questions(access_key):
+    survey = helpers.get_survey_data(access_key)
+    restrict_message = helpers.restrict_edit_access(survey, False)
+    if restrict_message:
+        flask.flash(restrict_message)
+        return list_surveys()
+
+    questions = flask.request.get_json(force=True)
+    helpers.set_questions(survey['survey_id'], questions)
+    return flask.jsonify({'success': True})
+
+
+@blueprint.route('/1/surveys/<access_key>', methods=['DELETE'])
+def delete_survey(access_key):
+    survey = helpers.get_survey_data(access_key)
+    restrict_message = helpers.restrict_edit_access(survey, True)
+    if restrict_message:
+        return flask.jsonify({'success': False, 'message': restrict_message})
+
+    helpers.delete_survey(survey['survey_id'])
+    return flask.jsonify({'success': True})
+
+
+@blueprint.route('/1/surveys/<access_key>/submit', methods=['POST'])
+def submit(access_key):
+    def error(message):
+        return flask.jsonify({'success': False, 'message': message})
+
+    survey = helpers.get_survey_data(access_key)
+    restrict_message = helpers.restrict_take_access(survey)
+    if restrict_message: return error(restrict_message)
+    post_data = flask.request.get_json(force=True)
+    if survey['auth']:
+        user_id = helpers.get_user_id(flask.session['username'])
+        auth_info = get_member_data(user_id, ['uid', 'birthday'])
+        authorized = post_data['uid'] == auth_info['uid'] and post_data['birthday'] == auth_info['birthday'].strftime(
+            helpers.YYYY_MM_DD)
+        if not authorized: return error('Incorrect UID or birthday')
+    responses = post_data['responses']
+    expected_question_ids = helpers.get_question_ids(survey['survey_id'])
+    if [response['question']
+            for response in responses] != expected_question_ids:
+        return error('Survey questions have changed')
+
+    question_types = helpers.get_question_types()
+    response_jsons = []
+    for response in responses:
+        question_id = response['question']
+        value = response['response']
+        type_id = helpers.get_question_type(question_id)
+        if type_id == question_types['Dropdown']:
+            choice = helpers.get_choice(question_id, value)
+            if choice is None: return error('Invalid choice for dropdown')
+            response_json = choice['choice_id']
+        elif type_id == question_types['Elected position']:
+            if type(value) != list:
+                return error('Invalid response to elected position')
+            response_json = []
+            for order_value in value:
+                if type(order_value) == str:
+                    if order_value == helpers.NO: response_json.append(None)
```

Comment:
why append None if choice is NO? just asking haha

Reply:
Just because I am not allowing arbitrary strings in the responses